File: CanzoniereWPF/Converters/DistanzaConverter.vb
Public Class DistanzaConverter
    Implements IValueConverter

    Public Function Convert(value As Object, targetType As Type, parameter As Object, culture As Globalization.CultureInfo) As Object Implements IValueConverter.Convert
        Dim d = DirectCast(value, Distanza)
        If targetType = GetType(Single) Then
            Return CType(d, Single)
        ElseIf targetType = GetType(Double) Then
            Return CType(d, Double)
        Else
            Throw New InvalidOperationException("Il tipo di destinazione deve essere un numerico")
        End If
    End Function

    Public Function ConvertBack(value As Object, targetType As Type, parameter As Object, culture As Globalization.CultureInfo) As Object Implements IValueConverter.ConvertBack
        Dim v As Single
        If IsNumeric(value) Then
            Return CType(value, Distanza)
        ElseIf Single.TryParse(value.ToString, v) Then
            Return CType(v, Distanza)
        Else
            Return DependencyProperty.UnsetValue
        End If
    End Function
End Class

File: KantaroWpf/Converters/DistanzaConverter.cs
using System;
using System.Globalization;
using System.Windows.Data;
using System.Windows;

namespace Enzo.Music.KantaroWpf.Converters;

public partial class DistanzaConverter : IValueConverter
{
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        Distanza d = (Distanza)value;
        if (targetType == typeof(float))
        {
            return (object)(float)d;
        }
        else if (targetType == typeof(double))
        {
            return (object)(double)d;
        }
        else
        {
            throw new InvalidOperationException("Il tipo di destinazione deve essere un numerico");
        }
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        if (Microsoft.VisualBasic.Information.IsNumeric(value))
        {
            return (Distanza)value;
        }
        else if (float.TryParse(value.ToString(), out float v))
        {
            return (Distanza)v;
        }
        else
        {
            return DependencyProperty.UnsetValue;
        }
    }
}
